packages/playwright/src/runner/watchMode.ts

```diff
@@ -19,25 +19,77 @@ import { createGuid, getPackageManagerExecCommand, ManualPromise } from 'playwri
 import type { FullConfigInternal, FullProjectInternal } from '../common/config';
 import { createFileMatcher, createFileMatcherFromArguments } from '../util';
 import type { Matcher } from '../util';
-import { TestRun, createTaskRunnerForWatch, createTaskRunnerForWatchSetup } from './tasks';
 import { buildProjectsClosure, filterProjects } from './projectUtils';
-import { collectAffectedTestFiles } from '../transform/compilationCache';
 import type { FullResult } from '../../types/testReporter';
-import { chokidar } from '../utilsBundle';
-import type { FSWatcher as CFSWatcher } from 'chokidar';
 import { colors } from 'playwright-core/lib/utilsBundle';
 import { enquirer } from '../utilsBundle';
 import { separator } from '../reporters/base';
 import { PlaywrightServer } from 'playwright-core/lib/remote/playwrightServer';
-import ListReporter from '../reporters/list';
+import { TestServerDispatcher } from './testServer';
+import { EventEmitter } from 'stream';
+import { type TestServerSocket, TestServerConnection } from '../isomorphic/testServerConnection';
 
-class FSWatcher {
-  private _dirtyTestFiles = new Map<FullProjectInternal, Set<string>>();
-  private _notifyDirtyFiles: (() => void) | undefined;
-  private _watcher: CFSWatcher | undefined;
-  private _timer: NodeJS.Timeout | undefined;
+class InMemoryServerSocket extends EventEmitter implements TestServerSocket {
+  constructor(public readonly send: (data: any) => void, public readonly close: () => void = () => {}) {
+    super();
+  }
+
+  addEventListener(event: string, listener: (e: any) => void) {
+    this.addListener(event, listener);
+  }
+}
+
+export async function runWatchModeLoop(config: FullConfigInternal): Promise<FullResult['status']> {
```

Comment:
I think we should not pass `FullConfigInternal` to this method. Instead, the `TestServer` should load the config, and we should operate on some kind of options, maybe even on `ConfigCLIOverrides`. See how `program.ts` calls `testServer.runUIMode` instead of loading the config.

Reply:
done!